### PR TITLE
Make REST communication with action containers more robust

### PR DESCRIPTION
**Pull request by sven-lange-last**
On systems with high load, POSTing the action container `/init` endpoint occasionally fails with `NoRouteToHostException`. Retry if this exception occurs.

## Description
* So far, the HTTP client retried when the connection to an action container failed with `HttpHostConnectException`. This is safe because no data has been transmitted yet when this exception occurs. Usually, this exception will occur if the action container startup is slow such that the target socket is not listening yet.
* This change adds `NoRouteToHostException` as retryable exception. Root cause for this exception is high system load. It is safe to retry after this exception occurred because it implies that no data has been transmitted yet.
* Needed to restructure the code to add the second retryable exception.
* Lowered the time between retries from 100 ms to 10 ms because this shorter time should be sufficient to have routing available / the target socket listening.
* Added logging so that retries can be seen in system logs.

## Related issue and scope

- [ ] I opened an issue to propose and discuss this change (#????)

## My changes affect the following components

- [ ] API
- [ ] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [x] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [ ] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes

- [x] Bug fix (generally a non-breaking change which closes an issue).
- [ ] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:

- [x] I signed an [Apache CLA](https://github.com/apache/incubator-openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/incubator-openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [ ] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [ ] I updated the documentation where necessary.

### CODE REVIEW

File: common/scala/src/main/scala/whisk/core/containerpool/HttpUtils.scala [hunk 1]
```diff
@@ -105,15 +112,33 @@ protected[core] class HttpUtils(hostname: String, timeout: FiniteDuration, maxRe
...
-      case Failure(t: HttpHostConnectException) if retry =>
-        if (timeoutMsec > 0) {
-          Thread sleep 100
-          val newTimeout = timeoutMsec - 100
-          execute(request, newTimeout, retry)
+      case Success(response) => response
+      case Failure(t: RetryableConnectionError) =>
+        if (retry) {
+          val sleepTime = 10.milliseconds
+          if (timeout.length > 0) {
```
Comment (markusthoemmes) [inline common/scala/src/main/scala/whisk/core/containerpool/HttpUtils.scala]: This should be `if(timeout > Duration.Zero)`

File: common/scala/src/main/scala/whisk/core/containerpool/HttpUtils.scala [hunk 2]
```diff
@@ -141,14 +166,15 @@ protected[core] class HttpUtils(hostname: String, timeout: FiniteDuration, maxRe
...
+  def post(host: String, port: Int, endPoint: String, content: JsValue)(implicit logging: Logging,
+                                                                        tid: TransactionId): (Int, Option[JsObject]) = {
     val connection = new HttpUtils(s"$host:$port", 90.seconds, 1.MB)
     val response = connection.post(endPoint, content, retry = true)
     connection.close()
     response match {
       case Right(r)                   => (r.statusCode, Try(r.entity.parseJson.asJsObject).toOption)
       case Left(NoResponseReceived()) => throw new IllegalStateException("no response from container")
-      case Left(Timeout())            => throw new java.util.concurrent.TimeoutException()
+      case Left(Timeout(t))           => throw new java.util.concurrent.TimeoutException()
```
Comment (markusthoemmes) [inline common/scala/src/main/scala/whisk/core/containerpool/HttpUtils.scala]: Could drop the value (i.e. use `Timeout(_)`)

File: common/scala/src/main/scala/whisk/core/containerpool/HttpUtils.scala [hunk 3]
```diff
@@ -68,20 +71,24 @@ protected[core] class HttpUtils(hostname: String, timeout: FiniteDuration, maxRe
...
 
-    execute(request, timeout.toMillis.toInt, retry)
+    execute(request, timeout, retry)
   }
 
-  private def execute(request: HttpRequestBase,
-                      timeoutMsec: Integer,
-                      retry: Boolean): Either[ContainerHttpError, ContainerResponse] = {
+  // Used internally to wrap all exceptions for which the request can be retried
+  private case class RetryableConnectionError(t: Throwable) extends Exception(t) with NoStackTrace
```
Comment (markusthoemmes) [inline common/scala/src/main/scala/whisk/core/containerpool/HttpUtils.scala]: Nice!

File: common/scala/src/main/scala/whisk/core/containerpool/HttpUtils.scala [hunk 4]
```diff
@@ -105,15 +112,33 @@ protected[core] class HttpUtils(hostname: String, timeout: FiniteDuration, maxRe
...
+      case t: NoRouteToHostException => Failure(RetryableConnectionError(t))
     } match {
-      case Success(r) => r
-      case Failure(t: HttpHostConnectException) if retry =>
-        if (timeoutMsec > 0) {
-          Thread sleep 100
-          val newTimeout = timeoutMsec - 100
-          execute(request, newTimeout, retry)
+      case Success(response) => response
+      case Failure(t: RetryableConnectionError) =>
```
Comment (markusthoemmes) [inline common/scala/src/main/scala/whisk/core/containerpool/HttpUtils.scala]: Could add the `if retry` as a guard here, the `else` branch and the more general `Left` case are equivalent.

File: common/scala/src/main/scala/whisk/core/containerpool/HttpUtils.scala (hunk 1 above)
Comment (sven-lange-last) [reply common/scala/src/main/scala/whisk/core/containerpool/HttpUtils.scala]: Makes perfect sense. I will change.

File: common/scala/src/main/scala/whisk/core/containerpool/HttpUtils.scala (hunk 2 above)
Comment (sven-lange-last) [reply common/scala/src/main/scala/whisk/core/containerpool/HttpUtils.scala]: Good catch. Done.

**Pull request merged**